text/0224-experimental-semantic-conventions.md

```diff
@@ -0,0 +1,117 @@
+A procedure will be established by the technical committee (TC), or an entity designated by the TC, for registration of new semantic convention attributes.
+Any attributes not intended strictly for private use SHOULD be registered in the attribute dictionary.
+
+Instrumentation authors SHOULD assume that any attribute they create may become standardized, public, commonly deployed, or usable across multiple OpenTelemetry clients, distributions, or receivers.
+They SHOULD choose attribute keys that are descriptive, relevant, and that they have reason to believe are not currently in use or already proposed.
+Before creating a new attribute, the attribute author SHOULD review the attribute dictionary to see if an attribute already exists which fits the intended use case.
+If such a registered attribute exists, then it SHOULD be preferred over the creation of new attributes unless there is sufficient reason to create a new attribute.
+Any new attributes SHOULD be registered in the attribute dictionary before an instrumentation using the attribute is released.
+If an instrumentation uses an attribute not included in the semantic conventions it is recommended that the attribute not be collected by default unless it is fundamental to the instrumentation.
+For example, an instrumentation which instruments a component not yet addressed by semantic conventions.
+
+An entry MUST NOT be refused without credible reason to believe that registration would be harmful.
+Note that a disagreement about technical details are not considered grounds to refuse listing in the attribute dictionary.
+Some examples of reasonable grounds for refusal: frivolous use of the dictionary such as registering nonsense words or inside jokes, a credible security or privacy risk, conflict with another attribute or proposal for a new attribute, or that the proposal is sufficiently lacking in purpose, or misleading about its purpose, that it can be held to be a waste of time and effort.
+This list is non-exhaustive and final decision for refusal is at the discretion of the TC.
+
+Once an attribute is registered it MUST NOT receive any breaking changes.
+This document explicitly does NOT define a process for removal of a registered attribute from the attribute dictionary.
+Because absence of evidence does not constitute evidence of absence, there is no way to know if any registered attribute is in active use.
+If a registered attribute is deemed by the TC to be harmful in some way, it SHOULD be marked deprecated, sufficient reasoning for the deprecation should be included in the dictionary, and, if appropriate, a new attribute SHOULD be registered with a different name.
+A registered attribute MAY ONLY be removed if, at the discretion of the TC, it is determined to be harmful or offensive to the project, its contributors, users, or some other group.
```

Comment:
I suggest applying this requirement to stable attributes only

Reply:
I don't want to distinguish between stable and experimental attributes. Any attribute in use should be stable or we will be breaking users, and any registered attribute should be assumed to be in use as we have no way to prove otherwise.

Reply:
based on our experience in HTTP semconv, we renamed certain attributes a couple of times before we were happy with them.

having `http.method`, and `http.request.method`, and `http.request_method_canonical` meaning the same thing with slight variations in definition would be confusing and harmful. 

Also, based on this proposal semantic conventions cannot use attributes outside of the dictionary, which means that, however premature new semconv is, attributes get into the dictionary first, before prototyping/development of semconv is complete. It's unrealistic to expect the first version of an attribute to be the final one. 

So we either need
- a separate experimental dictionary
- an experimental level for attributes
- allow experimental semconv define their own attributes within that semconv and only move attributes into the dictionary once semconv is in the feature-freeze/stable

Reply:
> I don't want to distinguish between stable and experimental attributes. Any attribute in use should be stable or we will be breaking users, and any registered attribute should be assumed to be in use as we have no way to prove otherwise.

I agree with @lmolkova here. Regarding breaking users by breaking attributes: how would it be different if we deprecate attributes in favor of introducing new ones? Effectively, this is a breaking change as well, because instrumentations will stop populating the deprecated attributes at some point in time (otherwise they would need to populate redundant attributes forever). 

Also, as I wrote below, I think we need a way to indicate for user how much they can rely on the stability of attributes. Some attributes can be rather stable (where breaking changes are very unlikely) other are intentionally not stable as they have been created for "experimentation purposes". But, how would (non-expert) semconv users know what applies to which attribute?

That's why I'm in favor of a stability level for attributes, too. 

In ECS we have a stability level per namespace and per field level. [See this example](https://github.com/elastic/ecs/blob/0525ea01e36ddd7315ef1153627dfba7eb3fdf70/schemas/process.yml#L341).

Reply:
> Also, as I wrote below, I think we need a way to indicate for user how much they can rely on the stability of attributes.

Great point! It means that stable -> deprecated is not a good lifecycle for the attribute, we still need the experimentation stage to communicate to users if they can really depend on the attribute

Reply:
>I agree with @lmolkova here. Regarding breaking users by breaking attributes: how would it be different if we deprecate attributes in favor of introducing new ones? Effectively, this is a breaking change as well, because instrumentations will stop populating the deprecated attributes at some point in time (otherwise they would need to populate redundant attributes forever).

It is different because when you are interpreting your data you can still look up an old attribute in the dictionary and find out its meaning. If the definition is changed or it is renamed without keeping the old name around somewhere, this information is simply lost.

I think an attribute-specific experimental phase is reasonable, but I think it would be better if previous versions were kept around in some way so that older data can be interpreted. What would you think if we had an attribute-specific version? Anyone can register an experimental `0.x` attribute, but only semconv maintainers can promote attributes to `1.x`?

Reply:
yes that is the idea. if that isn't clear from the text I can update it?